core/src/main/java/org/ldaptive/LdapUtils.java

```diff
@@ -195,6 +199,37 @@ public static String percentEncode(final String value)
   }
 
 
+  /**
+   * Converts all characters <= 0x1F and 0x7F to percent encoded hex.
+   *
+   * @param  value  to encode control characters in
+   *
+   * @return  string with percent encoded hex characters
+   */
+  public static String percentEncodeControlChars(final String value)
+  {
+    if (value != null) {
+      final Matcher m = CNTRL_PATTERN.matcher(value);
+      if (m.find()) {
+        final StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < value.length(); i++) {
```

Comment:
You should be able to implement this more compactly and efficiently using `Matcher#appendReplacement` and `Matcher#appendTail`. See the following javadoc for an example:

https://docs.oracle.com/javase/8/docs/api/java/util/regex/Matcher.html#appendReplacement-java.lang.StringBuffer-java.lang.String-

In particular, you shouldn't have to explicitly make assumptions about the specific characters in the class that matched like on line 218.

Reply:
I believe this is equivalent:

      final StringBuffer sb = new StringBuffer();
      while (m.find()) {
        // Each control character is one byte
        final byte[] bytes = new byte[1];
        // CheckStyle:MagicNumber OFF
        bytes[0] = (byte) (m.group().charAt(0) & 0x7F);
        // CheckStyle:MagicNumber ON
        m.appendReplacement(sb, new String(hexEncode(bytes)));
      }
      m.appendTail(sb);

Reply:
I explored that solution, but since it creates a new StringBuffer for *every* toString invocation I went a different way. I only wanted to pay the price if a control character was present.

Reply:
Ok, just wanted to make note of it.